aten/src/ATen/native/cuda/Unique.cu

```diff
@@ -15,148 +16,211 @@ namespace native{
+) {
+
+  /**
+    * The idea for implementing this is basically the same as unique.
+    * For unique_dim, we are taking the unique with respect to a index
+    * tensor, but during the processes, we override the compare and equal
+    * operator by checking the data underlying it instead. After the
+    * algorithm, we would use index_select to map the resulting indicies
+    * to the result on the actual data.
+    */
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+
+  int64_t num_inp = self.size(dim);
+  auto options = self.options().dtype(kLong);
+  Tensor input_flat = self.transpose(dim, 0).contiguous().view({num_inp, -1});
+  int64_t n = input_flat.size(1);
+  scalar_t *input_flat_ptr = input_flat.data<scalar_t>();
+
+  Tensor indices = at::arange(0, num_inp, options);
+  int64_t *indices_data = indices.data<int64_t>();
+  thrust::sort(policy, indices_data, indices_data + num_inp,
+    [=] __device__ (int64_t a, int64_t b) -> bool {
+      for (int64_t i = 0; i < n; ++i) {
+        scalar_t lhs = input_flat_ptr[i + a * n];
+        scalar_t rhs = input_flat_ptr[i + b * n];
+        if (lhs < rhs) {
+          return true;
+        } else if (lhs > rhs) {
+          return false;
         }
-        return true;
-      });
-    input_sorted_indices.resize_(last - input_sorted_indices_ptr);
-    Tensor output = input_sorted.index_select(0, input_sorted_indices);
-
-    // reshape back
-    auto new_sizes = std::vector<int64_t>(orig_sizes);
-    new_sizes[0] = -1;
-    output = output.view(new_sizes);
-    output = output.transpose(0, dim);
-
-    // calculate inverse indices
-    Tensor inverse_indices = at::empty({0}, self.type().toScalarType(kLong));
-    if (return_inverse) {
-      int64_t size = self.size(dim);
-      inverse_indices.resize_(size);
-      Tensor mask = at::empty(input_sorted.size(0), self.type().toScalarType(kLong));
-      mask[0] = 1;
-      for (int i = 0; i < input_sorted.size(0) - 1; ++i) {
-        if (!at::equal(input_sorted[i], input_sorted[i+1])) {
-          mask[i+1] = 1;
-        } else {
-          mask[i+1] = 0;
+      }
+      return false;
+    }
+  );
+
+  Tensor inverse_indices, counts;
+  int64_t num_out;
+  std::tie(inverse_indices, counts, num_out) = compute_unique(
+    policy, indices_data, num_inp, indices_data,
+    return_inverse, return_counts, options,
+    [=] __device__ (int64_t a, int64_t b) -> bool {
+      for (int64_t i = 0; i < n; ++i) {
+        scalar_t lhs = input_flat_ptr[i + a * n];
+        scalar_t rhs = input_flat_ptr[i + b * n];
+        if (lhs != rhs) {
+          return false;
         }
       }
-
-      Tensor imask = at::cumsum(mask, 0) - 1;
-      for (int i = 0; i < indices.size(0); ++i) {
-        inverse_indices[indices[i]] = imask[i];
+      return true;
+    },
+    [=] __device__ (int64_t a, int64_t b) -> int64_t {
```

Comment:
Nit: there's no real reason for this to return int64_t rather than bool, right? thrust::not_equal_to that is used in the non-dim case returns bool.

Reply:
@ngimel Yes, you are right. C++ would automatically convert `true` to `1`. But does changing `int64_t` to `bool` here improve anything? If not, I will leave it as `int64_t`.

Reply:
You are right, it won't improve anything, it will just be similar to `equal` used a few lines above, and thrust functors used in the non-dim case. Feel free to leave as is.